executor/include/executor/best_fit.hpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2014-, Open Perception, Inc.
+ *  Author(s): Shrijit Singh <shrijitsingh99@gmail.com>
+ *
+ */
+
+#pragma once
+
+#include <executor/type_trait.h>
+
+#include <executor/default/cuda_executor.hpp>
+#include <executor/default/inline_executor.hpp>
+#include <executor/default/omp_executor.hpp>
+#include <executor/default/sse_executor.hpp>
+
+namespace executor {
+
+static const auto best_fit_executors =
+    std::make_tuple(executor::omp_executor<>{}, executor::sse_executor<>{},
+                    executor::inline_executor<>{});
+
+struct executor_runtime_checks {
+  template <typename Executor, typename executor::InstanceOf<
+                                   Executor, executor::inline_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::InstanceOf<
+                                   Executor, executor::sse_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::InstanceOf<
+                                   Executor, executor::omp_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::InstanceOf<
+                                   Executor, executor::cuda_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+};
+
+namespace detail {
+
+template <typename Function, typename Executor, typename = void>
+bool execute(Function&& f, Executor& exec) {
+  return false;
+}
+
+template <typename Function, template <typename...> class Executor, typename... Properties,
+          typename std::enable_if<executor::is_executor_available_v<Executor>,
+                                  int>::type = 0>
+bool execute(Function&& f, Executor<Properties...>& exec) {
+  f(exec);
+  return true;
+}
+
+}  // namespace detail
+
+template <typename RuntimeChecks = executor_runtime_checks, typename Function,
+          typename... SupportedExecutors>
+void enable_exec_with_priority(
+    Function f, std::tuple<SupportedExecutors...> supported_execs) {
+  static_assert(std::is_base_of<executor_runtime_checks, RuntimeChecks>::value,
```

Comment:
Why is this required? Why not just accept a functor (or lambda) and call that with the executor?

Reply:
I am not quite sure what you are referring to by this?
Do you mean `static_assert(std::is_base_of<executor_runtime_checks, RuntimeChecks>::value,
`?

Reply:
I meant why do we need a struct with a `check` static function, why not any functor? Let the user wrap things in a lambda anyhow (such as [overload idiom](https://dev.to/tmr232/that-overloaded-trick-overloading-lambdas-in-c17))